js/boaexchange.js

```diff
@@ -0,0 +1,850 @@
+        return orders;
+    }
+
+    async fetchDepositAddress (code, params = {}) {
+        await this.loadMarkets ();
+        const currency = this.currency (code);
+        const request = {
+            'label': currency['id'],
+        };
+        const response = await this.v1GetAddressesLabel (this.extend (request, params));
+        let address = this.safeString (response['data']['address'], 'address');
+        if (!address) {
+            throw new AddressPending (this.id + ' the address for ' + code + ' is being generated (pending, not ready yet, retry again later)');
+        }
+        this.checkAddress (address);
+        return {
+            'currency': code,
+            'address': address,
+            'tag': undefined,
+            'info': response,
+        };
+    }
+
+    async withdraw (code, amount, address, tag = undefined, params = {}) {
+        this.checkAddress (address);
+        await this.loadMarkets ();
+        let currency = this.currency (code);
+        let request = {
+            'label': currency['id'],
+            'amount': amount,
+            'address': address,
+        };
+        let response = await this.v1GetAccountWithdraw (this.extend (request, params));
+        let id = undefined;
+        if ('data' in response) {
+            if ('withdraw' in response['data'])
+                id = response['data']['withdraw']['id'];
+        }
+        return {
+            'info': response,
+            'id': id,
+        };
+    }
+
+    nonce () {
+        return this.seconds ();
+    }
+
+    sign (path, api = 'public', method = 'GET', params = {}, headers = undefined, body = undefined) {
+        let url = this.implodeParams (this.urls['api'][api], {
+            'hostname': this.hostname,
+        }) + '/v1/';
+        url += this.implodeParams (path, params);
+        params['limit'] = 500;
+        url += '?' + this.urlencode (params);
+        this.checkRequiredCredentials ();
+        let nonce = '' + this.nonce ();
+        let signature = this.hmac (this.encode (nonce), this.encode (this.secret), 'sha256');
+        headers = {
+            'X-BOA-ENCRYPTED': signature,
+            'X-BOA-KEY': this.apiKey,
+            'X-BOA-NONCE': nonce,
+            'Content-Type': 'application/json',
+        };
+        return { 'url': url, 'method': method, 'body': body, 'headers': headers };
+    }
+
+    handleErrors (code, reason, url, method, headers, body, response) {
+        if (body[0] === '{') {
+            let data = this.safeValue (response, 'data');
+            let errors = this.safeValue (response, 'errors');
+            if (errors !== undefined)
+                throw new ExchangeError (this.id + ': (URL: ' + url + ') an error occoured: ' + this.json (errors));
+            if (data === undefined)
+                throw new ExchangeError (this.id + ': malformed response: ' + this.json (response));
+        }
+    }
+
+    appendTimezoneParse8601 (x) {
```

Comment:
Ok, first of, thx for the contribution! It's much cleaner than what we get on average, but still has a few things that we will change in this file to merge it. I'd suggest to start by cleaning out the unnecessary copypasted methods like this one, `appendTimezoneParse8601`, `getOrderIdField`. We also need to check the endpoints lists in the `describe()` method, because it's not clear how balances can be deleted: https://github.com/ccxt/ccxt/pull/4813/files#diff-cdb3157a354625d9b6af59f245b8889eR108 ) Seems like that list of endpoints is copypasted multiple times, but we should be picky in what goes where, namely, the endpoints should go under corresponding http methods (get/post/delete).

Reply:
Yep, hadn't yet committed all my changes, fixed in next commits